Readme.md

# express-rate

express-rate is a express.js route specific middleware that monitors the rate at which specific routes get requested and allows you to limit a specific requester. This can be a basic prevention of DOS attacks, or simply used as a Twitter-like rate-limiter on your API. Comes with support for an in memory store, and a Redis store backed by [node_redis](http://github.com/mranney/node_redis).

This is an alpha release. 

## Installation

	  $ npm install express-rate

Note: this module requires redis > 2.1.3 because of changes in the behavior of the [expire](http://redis.io/topics/expire) action.


## Options:

- *handler*

```js
new rate.Redis.RedisRateHandler({client: redis.createClient()})
```

  The rate handler that stores the routes for the rate. The built in handlers are for Redis (recommended) 
  and Memory (for testing or single webserver clusters - use new rate.Memory.MemoryRateHandler() ).

- *interval*

```js
1
```

  The interval of time in *seconds* that the middleware tracks your route. For interval=1, it will track 
  the amount of requests hitting your route per 1 second. For value=10, it will track requests per 10 seconds.

- *limit*

```js
0
```

  For any non-zero values of limit, rate-limiting will be turned on. For limit=10 and interval=1, the system
  will only allow 10 requests per second for a specific requester on this specific route. 
  For limit=100 and interval=10, the system will allow 100 requests per 10 seconds for this specific requester on this specific route.


- *getRemoteKey*

```js
function (req) {
  return req.connection.remoteAddress;
}
```

  A function that allows you to switch how rate identifies a specific requester. By default, it is their IP. You can override it with a function that sets it based on the authenticated user, or on a API key that the user is posting. See below in usage for an example. 

- *getRouteKey*

```js
function (req) {
  return req.route.method + ':' + req.route.regexp;
}
```

  A function that allows you to switch how rate identifies a route. By fefault, its the method:route_regex

- *setHeaders*

```js
true
```

  Whether you want to set rate limiting headers or not.

- *setHeadersHandler*

```js
function (req, res, rate, limit, resetTime) {

    var remaining = limit - rate;
    if (remaining < 0) {
        remaining = 0;
    }

    // follows Twitter's rate limiting scheme and header notation
    // https://dev.twitter.com/docs/rate-limiting/faq#info
    res.setHeader('X-RateLimit-Limit', limit);
    res.setHeader('X-RateLimit-Remaining', remaining);
    res.setHeader('X-RateLimit-Reset', resetTime);
}
```

  A function that is called to cutomize the request and its headers to give the requester an idea of how rate-limited it is.

- *onLimitReached*

```js
function (req, res, rate, limit, resetTime, next) {
            
    // HTTP code 420 from http://mehack.com/inventing-a-http-response-code-aka-seriously
    res.json({error: 'Rate limit exceeded. Check headers for limit information.'}, {status: 420});
}
```

  A function that is called when a route's rate limit is reached. This will allow you to customize what you
  return back to the requester when they can no longer have access to this route.



## Usage

Below is code demonstrating use and customization of most of the use cases of rate:

```js

var rate            = require('../lib/rate'),
  redis     = require('redis'),
  client      = redis.createClient();

var app = require('express').createServer();

// create the rate redis handler
var redisHandler = new rate.Redis.RedisRateHandler({client: client});



// 
// Monitor rate of a route
//

// options: handler is the mechanism where rate information is kept
// interval is the interval on which we are listening for the rate. Here we're counting requests per 1 second.
// If we switch interval to 10, then we are counting how many requests this route is handling per 10 seconds.
var simpleMiddleware = rate.middleware({handler: redisHandler, interval: 1});

// apply the rate middleware to the / route
app.get('/', simpleMiddleware, function(req, res, next){
  res.send('I am being rate monitored at requests per 1 second.')
});

//
// Monitor the rate of this route, and report the RPS of **everyone** hitting this route
//
var monitorMiddleware1 = rate.middleware({handler: redisHandler, interval: 1});

// apply the rate middleware to the / route
app.get('/route-rate', monitorMiddleware1, function(req, res, next){
  var rateKey = req.route.method + ':' + req.route.regexp;
  redisHandler.getRate(rateKey, null, next, function (rate) {
    res.send('This route is being requested at ' + rate + ' requests per 1 second.');
  });
});

//
// Monitor the rate of this route, and report the RPS of **only you** hitting this route
//
var monitorMiddleware2 = rate.middleware({handler: redisHandler, interval: 1});

// apply the rate middleware to the / route
app.get('/me-rate', monitorMiddleware2, function(req, res, next){
  var rateKey = req.route.method + ':' + req.route.regexp;
  var remoteKey = req.connection.remoteAddress;
  redisHandler.getRate(rateKey, remoteKey, next, function (rate) {
    res.send(remoteKey + ' is requesting this route at ' + rate + ' requests per 1 second.');
  });
});

//
// Monitor and rate-limit this route
//

// this middleware will only allow 2 requests for every 4 seconds from a specific user
var limiterMiddleware = rate.middleware({handler: redisHandler, interval: 4, limit: 2});

app.get('/limited', limiterMiddleware, function(req, res){
  res.send('You can only request me twice every 4 seconds!');
});


// 
// Key requesters by their API key instead of IP
//

var apiKeyMiddleware = rate.middleware(
  {handler: redisHandler, 
   limit: 10,
   interval: 2, 
   getRemoteKey: function (req) {
    return req.params.api_key;
   }
});

app.get('/api/:api_key/operation', apiKeyMiddleware, function(req, res){
  res.send('API key ' + req.params.api_key + ' only has ' + res.getHeader('x-ratelimit-remaining') + ' remaining requests.');
});

// 
// Customize the headers sent back during rate limiting and the message returned
//

var headersMiddleware = rate.middleware(
  {handler: redisHandler, 
   limit: 5,
   interval: 5, 
   setHeadersHandler: function (req, res, rate, limit, resetTime) {

            var remaining = limit - rate;

            if (remaining < 0) {
                remaining = 0;
            }

            // follows Twitter's rate limiting scheme and header notation
            // https://dev.twitter.com/docs/rate-limiting/faq#info
            res.setHeader('X-RateLimit-Limit', limit);
            res.setHeader('X-RateLimit-Remaining', remaining);
            res.setHeader('X-RateLimit-Reset', resetTime);
        },

        onLimitReached: function (req, res, rate, limit, resetTime, next) {
            
            // HTTP code 420 from http://mehack.com/inventing-a-http-response-code-aka-seriously
            res.json({error: 'Rate limit exceeded. Check headers for limit information.'}, {status: 420});
        }
});

app.get('/headers', headersMiddleware, function(req, res){
  var text = 'X-RateLimit-Limit :' + res.getHeader('X-RateLimit-Limit') + '              ' + 
    'X-RateLimit-Remaining :' + res.getHeader('X-RateLimit-Remaining') + '              ' + 
    'X-RateLimit-Reset :' + res.getHeader('X-RateLimit-Reset');
  res.send(text);
});

app.listen(3000);

```

## TO-DOs

  - Support multiple intervals
  - Fix redis tests to work all together

## Testing

Get the testing requirements:

    $ npm install expresso -g
    $ npm install should

Go to the root project directory and run:

    $ make test

Note: redis tests (/test/redis.tests.js) work seperately but not together.

History.md
0.0.1 / 2011-10-28 
==================

  * Initial commit
  * Supports basic route and identity rate monitoring and limiting

test/redis.test.js

var common          = require('./common.rate');
    rate            = require('../lib/rate'),
    redis           = require('redis'),
    client          = redis.createClient();

var onFinished = function () {
    client.end();
};

module.exports = {


    //
    // TODO: I have not figured out how tu run these tests in parallel or in serial (--serial flag) because of the
    // shared REDIS resource, if run one at a time, each of them pass for me, but not together
    // 

    /*
    'baseline: server can be tested' : function () {
        common['baseline: server can be tested']();
    },


    'server records rate': function (done) {
        var handler = new rate.Redis.RedisRateHandler({client: client});
        common['server records rate'](handler, onFinished);
    },

    'Reset after interval': function (done) {
        var handler = new rate.Redis.RedisRateHandler({client: client});
        common['Reset after interval'](handler, onFinished);
    },
    */

    'Routes can be rate limited, reallowed, and have proper headers': function (done) {
        var handler = new rate.Redis.RedisRateHandler({client: client});
        common['Routes can be rate limited, reallowed, and have proper headers'](handler, onFinished);
    }

};

test/common.rate.js


var express         = require('express'),
    assert          = require('assert'),
    should          = require('should'),

    rate            = require('../lib/rate'),
    memory          = require('../lib/memory');

module.exports = {

    routeKey: 'get:/^\\/\\/?$/i',
	
	'baseline: server can be tested' : function () {
		
		var server = express.createServer();

        var responseBody = 'works';

        server.get('/', function (req, res) {
            res.send(responseBody);
        });

        assert.response(server,
            { url: '/' },
            { body: responseBody });
	},

	'server records rate' : function (handler, done) {

        var self = this;

        handler.reset(module.exports.routeKey, null, null, function () {

    		var middleware = rate.middleware({handler: handler, interval: 2});
    		self.performRandomRequests(middleware, Math.random() * 100, function (responseNumber, requestsToPerform) {

    			
    			if (responseNumber === requestsToPerform) {

                    handler.getRate(module.exports.routeKey, null, null, function (rate) {
                    	
                        assert.eql(rate, requestsToPerform, 'Rate of received from middleware does not match sent.');
                        
                        done();
                    });
                }

    		});

        });
	},


	'Reset after interval': function (handler, done) {

        var self = this;

        handler.reset(module.exports.routeKey, null, null, function () {

    		var secondsIncrement = 1;
    		var resetTime = (new Date()).getTime() + (secondsIncrement * 1000);
    		var middleware = rate.middleware({handler: handler, interval: secondsIncrement});
    		var completed = false;
    		var inflections = 0;
    		var lastRateCount = null;

    		self.performRandomRequests(middleware, Math.random() * 15000, function (responseNumber, requestsToPerform) {

    			handler.getRate(module.exports.routeKey, null, null, function (rate) {
                    
                    if (lastRateCount && rate < lastRateCount)
                    	inflections += 1;
                    
                    lastRateCount = rate;
                       
                    if (responseNumber === requestsToPerform) {
                    	assert.eql((inflections > 0), true, 'There was no inflections, meaning the rate was never reversed, but was monotonically increasing.');
                        
                        done();
                    }
                });

    		});
        });
	},

	'Routes can be rate limited, reallowed, and have proper headers': function (handler, done) {
		
        var self = this;

        handler.reset(module.exports.routeKey, null, null, function () {

    		var server = express.createServer();

            var responseBody = 'works';
            var errorJson = 'limited';

            var limit = 10;
            var middleware = rate.middleware({handler: handler, interval: 5, limit: limit,
            	onLimitReached: function (req, res, rate, limit, resetTime, next) {
            
    		        // HTTP code 420 from http://mehack.com/inventing-a-http-response-code-aka-seriously
    		        res.send(errorJson, {status: 420});
    		    }
            });

            server.get('/', middleware,
                function (req, res) {
                    res.send(responseBody);
                }
            );

            var responsesReceived = 0;
            var stop = false;
            var limitedRequests = 0;
            var lastCheck = false;


            var makeRequest = function (response) {
                
                	assert.response(server,
                        { url: '/' },
                        response, 
                        function (res) {
                            responsesReceived += 1;

                            assert.isDefined(res.headers['X-RateLimit-Limit'.toLowerCase()]);
                            assert.isDefined(res.headers['X-RateLimit-Remaining'.toLowerCase()]);
                            assert.isDefined(res.headers['X-RateLimit-Reset'.toLowerCase()]);

                            // check the headers
                            assert.eql(res.headers['X-RateLimit-Limit'.toLowerCase()], limit, 'Limit does not match the header limit returned from server.');

                            if(res.body === errorJson) {
                            	limitedRequests += 1;
                            }
                            		
                            if (parseInt(res.headers['X-RateLimit-Remaining'.toLowerCase()]) > 0) {



                            	if (!lastCheck) {
                            		makeRequest({body: responseBody});
        	                    } else {

                                    done();
                                }
                            	
                             } else {

                             	if (limitedRequests > 5) {
                             		
                        			var currentTime = (new Date()).getTime();
                        			var sleepForMs = parseInt(res.headers['X-RateLimit-Reset'.toLowerCase()]) - currentTime;
                                    // TODO: bug, have to add an extra second for differences between redis time and local time
                                    // need to find correct way to do this
                        			sleepForMs += 1000;

                        			setTimeout(function () {

                        				lastCheck = true;
                        				makeRequest({body: responseBody});

                        			}, sleepForMs);

                             	} else {

                            		makeRequest({body: errorJson});
                            	}
                            }
                        }
                    );
                
            }

            makeRequest({body: responseBody});
        });
	},

	performRandomRequests: function (middleware, n, callback) {

		var server = express.createServer();

        var responseBody = 'works';

        
        server.get('/', middleware,
            function (req, res) {
                res.send(responseBody);
            }
        );

        var requestsToPerform = Math.ceil(n),
            responsesReceived = 0;

        for(var i = 0; i < requestsToPerform; i += 1) {
            
            assert.response(server,
                { url: '/' },
                { body: responseBody }, 
                function (res) {
                    responsesReceived += 1;
                    callback(responsesReceived, requestsToPerform);
                }
            );
        }
	}

}

test/memory.test.js

var common          = require('./common.rate');
    rate            = require('../lib/rate');

var onFinished = function () {
    
};

module.exports = {

    'baseline: server can be tested' : function () {
        common['baseline: server can be tested']();
    },

    'server records rate': function (done) {
        var handler = new rate.Memory.MemoryRateHandler();
        common['server records rate'](handler, onFinished);
    },

    'Reset after interval': function (done) {
        var handler = new rate.Memory.MemoryRateHandler();
        common['Reset after interval'](handler, onFinished);
    },

    'Routes can be rate limited, reallowed, and have proper headers': function (done) {
        var handler = new rate.Memory.MemoryRateHandler();
        common['Routes can be rate limited, reallowed, and have proper headers'](handler, onFinished);
    }
};

lib/rate.js

exports.Base             = require('./base');
exports.Memory           = require('./memory');
exports.Redis            = require('./redis');

// 
// DEFAULTS FOR OPTIONS
// extensibility point
//

exports.defaults = function () {

    return {

        handler: new exports.Memory.MemoryRateHandler(),

        getRemoteKey: function (req) {
            return req.connection.remoteAddress;
        },

        getRouteKey: function (req) {
            return req.route.method + ':' + req.route.regexp;
        },


        interval: 1,

        limit: 0,

        setHeaders: true,

        setHeadersHandler: function (req, res, rate, limit, resetTime) {

            var remaining = limit - rate;

            // remaining can be smaller than 0 because it just counts how many requests were made
            // and is incremented whether or not it is over the limit
            // we just won't allow the request to continue to the controller if its over the limit
            // we zero it here so it doesnt confuse the user

            // logic behind this is so for REDIS, we only have to make one increment request (which returns
            // the current count), with 
            // one roundtrip to the server. Otherwise, we would have to first check what the value is and
            // then increment if smaller (which is two roundtrips to the server)
            // if someone knows how to do conditionals in REDIS multi transactions, let me know!
            if (remaining < 0) {
                remaining = 0;
            }


            // follows Twitter's rate limiting scheme and header notation
            // https://dev.twitter.com/docs/rate-limiting/faq#info
            res.setHeader('X-RateLimit-Limit', limit);
            res.setHeader('X-RateLimit-Remaining', remaining);
            res.setHeader('X-RateLimit-Reset', resetTime);
        },

        onLimitReached: function (req, res, rate, limit, resetTime, next) {
            
            // HTTP code 420 from http://mehack.com/inventing-a-http-response-code-aka-seriously
            res.json({error: 'Rate limit exceeded. Check headers for limit information.'}, {status: 420});
        }

    };
};

//
// MIDDLEWARE
//

exports.middleware = function recordRate(options) {

    var defaults = exports.defaults();

    options = options || {};

    for (var prop in defaults) {
        if (options[prop] == null) options[prop] = defaults[prop];
    }

    return function recordRate(req, res, next) {

        var routeKey = options.getRouteKey(req),
            remoteKey = options.getRemoteKey(req),
            incrementCallback = null;

        // check if there is a limit set on this route, and reject request with headers if so
        if (options.limit) {

            // use the callback to get the rate
            incrementCallback = function (rate, resetTime) {

                if (options.setHeaders) {
                    options.setHeadersHandler(req, res, rate, options.limit, resetTime);
                }

                if (rate > options.limit) {

                    // we are officially over the limit
                    options.onLimitReached(req, res, rate, options.limit, resetTime, next);

                } else {

                    next();
                    
                }
            };
        }

        options.handler.increment(routeKey, remoteKey, options, next, incrementCallback);

        // if we don't have rate limiting turned on, then we don't care when incrementing the rate finishes
        // let's keep going with the request
        if (!incrementCallback)
            next();
    }
};


lib/redis.js

var sys             = require('sys'),
    base            = require('./base');

//
// DEFINE THE MEMORY RATE HANDLER
//

var RedisRateHandler = exports.RedisRateHandler = function (options) {
    this.options = options;
    this.client = this.options.client;
    this.initialized = !(typeof this.options.client == 'function');
};

// inherit from the base rate handler
sys.inherits(RedisRateHandler, base.BaseRateHandler);

RedisRateHandler.prototype.initialize = function () {
    if (typeof this.options.client == 'function') {
        this.client = this.options.client();
    }

    this.initialized = true;
};

RedisRateHandler.prototype._increment_key = function (key, increment, rate_options, next, callback) {

    if (!this.initialized)
        this.initialize();

    var currentTime = (new Date()).getTime(),
        ttl = currentTime + (rate_options.interval * 1000),
        rate = 1,
        self = this;

    this.client.multi()
        .incrby(key, increment)
        .ttl(key)
        .exec(function (err, replies) {
            
        if (err) next(err);

        currentTime = (new Date()).getTime();
        
        rate = replies[0];

        // if the key was just created now, we need to set an expiraton
        if(rate === increment) {
            ttl = currentTime + (rate_options.interval * 1000);
            self.client.expire(key, rate_options.interval);
        } else {
            ttl = currentTime + (replies[1] * 1000);
        }

        if (callback) callback(rate, ttl);
    });
};

RedisRateHandler.prototype._reset_key = function (key, increment, next, callback) {

    if (!this.initialized)
        this.initialize();

    this.client.del(key, function (err, reply) {
        if (err) next(err);
        if (callback) callback();
    });
};


RedisRateHandler.prototype.getRate = function (routeKey, remoteKey, next, callback) {
    
    if (!this.initialized)
        this.initialize();

    var fullKey = routeKey;
    if (remoteKey) {
        fullKey += ':' + remoteKey;
    }

    this.client.get(fullKey, function (err, rate) {
        if (err) next(err);

        if (rate)
            callback(rate);
        else
            callback(0);
    });   
};


lib/memory.js

var sys             = require('sys'),
    base            = require('./base');

//
// DEFINE THE MEMORY RATE HANDLER
//

var MemoryRateHandler = exports.MemoryRateHandler = function (options) {
    this.options = options;
    // stores the key value pairs corresponding with the rates of people using the site
    // TODO: flush out very old remote logs that have expired, REDIS will do that for you
    this.rates = {};
};

// inherit from the base rate handler
sys.inherits(MemoryRateHandler, base.BaseRateHandler);

MemoryRateHandler.prototype._increment_key = function (key, increment, rate_options, next, callback) {
    var remakeKey = true,
        currentTime = (new Date()).getTime(),
        ttl = currentTime,
        rate = 1;

    if (this.rates[key]) {
        
        var rateInfo = this.rates[key];

        if (currentTime < rateInfo.ttl) {
            // no need to remake the key
            remakeKey = false;
            // increment the counter
            rateInfo.value += increment;
            
            rate = rateInfo.value + increment;
            ttl = rateInfo.ttl;
        }
    }
    
    if (remakeKey) {

        ttl = currentTime + (rate_options.interval * 1000);
        this.rates[key] = {value: 1, ttl: ttl};

    }

    if (callback) callback(rate, ttl);
};

MemoryRateHandler.prototype._reset_key = function (key, increment, next, callback) {
    if (this.rates[key])
        delete this.rates[key];

    if (callback) callback();
};

MemoryRateHandler.prototype.getRate = function (routeKey, remoteKey, next, callback) {
    
    var fullKey = routeKey;
    if (remoteKey) {
        fullKey += ':' + remoteKey;
    }

    if (this.rates[fullKey]) {
        callback(this.rates[fullKey].value);
    } else {
        callback(0);
    }
        
};


examples/simple-redis.js
var rate            = require('../lib/rate'),
	redis			= require('redis'),
	client			= redis.createClient();

var app = require('express').createServer();

// create the rate redis handler
var redisHandler = new rate.Redis.RedisRateHandler({client: client});



// 
// Monitor rate of a route
//

// options: handler is the mechanism where rate information is kept
// interval is the interval on which we are listening for the rate. Here we're counting requests per 1 second.
// If we switch interval to 10, then we are counting how many requests this route is handling per 10 seconds.
var simpleMiddleware = rate.middleware({handler: redisHandler, interval: 1});

app.get('/', simpleMiddleware, function(req, res, next){
	res.send('I am being rate monitored at requests per 1 second.')
});

//
// Monitor the rate of this route, and report the RPS of **everyone** hitting this route
//
var monitorMiddleware1 = rate.middleware({handler: redisHandler, interval: 1});

app.get('/route-rate', monitorMiddleware1, function(req, res, next){
	var rateKey = req.route.method + ':' + req.route.regexp;
	redisHandler.getRate(rateKey, null, next, function (rate) {
		res.send('This route is being requested at ' + rate + ' requests per 1 second.');
	});
});

//
// Monitor the rate of this route, and report the RPS of **only you** hitting this route
//
var monitorMiddleware2 = rate.middleware({handler: redisHandler, interval: 1});

app.get('/me-rate', monitorMiddleware2, function(req, res, next){
	var rateKey = req.route.method + ':' + req.route.regexp;
	var remoteKey = req.connection.remoteAddress;
	redisHandler.getRate(rateKey, remoteKey, next, function (rate) {
		res.send(remoteKey + ' is requesting this route at ' + rate + ' requests per 1 second.');
	});
});

//
// Monitor and rate-limit this route
//

// this middleware will only allow 2 requests for every 4 seconds from a specific user
var limiterMiddleware = rate.middleware({handler: redisHandler, interval: 4, limit: 2});

app.get('/limited', limiterMiddleware, function(req, res){
  res.send('You can only request me twice every 4 seconds!');
});


// 
// Key requesters by their API key instead of IP
//

var apiKeyMiddleware = rate.middleware(
	{handler: redisHandler, 
	 limit: 10,
	 interval: 2, 
	 getRemoteKey: function (req) {
	 	return req.params.api_key;
	 }
});

app.get('/api/:api_key/operation', apiKeyMiddleware, function(req, res){
	res.send('API key ' + req.params.api_key + ' only has ' + res.getHeader('x-ratelimit-remaining') + ' remaining requests.');
});

// 
// Customize the headers sent back during rate limiting and the message returned
//

var headersMiddleware = rate.middleware(
	{handler: redisHandler, 
	 limit: 5,
	 interval: 5, 
	 setHeadersHandler: function (req, res, rate, limit, resetTime) {

            var remaining = limit - rate;

            if (remaining < 0) {
                remaining = 0;
            }

            // follows Twitter's rate limiting scheme and header notation
            // https://dev.twitter.com/docs/rate-limiting/faq#info
            res.setHeader('X-RateLimit-Limit', limit);
            res.setHeader('X-RateLimit-Remaining', remaining);
            res.setHeader('X-RateLimit-Reset', resetTime);
        },

        onLimitReached: function (req, res, rate, limit, resetTime, next) {
            
            // HTTP code 420 from http://mehack.com/inventing-a-http-response-code-aka-seriously
            res.json({error: 'Rate limit exceeded. Check headers for limit information.'}, {status: 420});
        }
});

app.get('/headers', headersMiddleware, function(req, res){
	var text = 'X-RateLimit-Limit :' + res.getHeader('X-RateLimit-Limit') + '              ' + 
		'X-RateLimit-Remaining :' + res.getHeader('X-RateLimit-Remaining') + '              ' + 
		'X-RateLimit-Reset :' + res.getHeader('X-RateLimit-Reset');
	res.send(text);
});

app.listen(3000);